torch/csrc/autograd/engine.cpp

```diff
@@ -392,6 +433,53 @@ auto Engine::evaluate_function(FunctionTask& task) -> void {
       is_ready = true;
     }
 
+    // Switches to child's stream for adding to InputBuffer
+    // If the child's stream is different from the parent's, 
+    // creates (once) an event and has the child wait on it
+    #ifdef WITH_CUDA
+      if (next.function->num_inputs() <= 0) {
+        throw std::runtime_error(std::string("child function with no inputs: " + next.function->name()));
+      }
+      
+      // Sets the child's device and stream
+      const int child_device = output.is_cuda() ? output.get_device() : - 1;
+      const TypeAndShape& child_metadata = next.function->input_metadata(next.input_nr);
+      auto child_stream = child_metadata.stream();
+
+      // Handles metadata mismatch
+      child_stream = 
+        (child_stream != nullptr 
+        && THCStream_device(child_stream) == child_device)
+        ? child_stream
+        : nullptr;
+
+      AutoGPUStream child_auto_gpu_stream{
+          child_device
+        , child_stream};
+
+      if (child_stream == nullptr) {
+        child_stream = auto_gpu_stream.reset_stream();
+      }
+
+      // Sync parent and child streams (only if they both exist)
+      // CPU -> CPU tasks are synchronized since there is only one CPU thread
+      // CPU -> GPU and GPU -> CPU tasks must implement their own synchronization
+      if (parent_stream != nullptr && child_stream != nullptr && parent_stream != child_stream) {
+        // Creates the event (once)
+        if(!has_event) {
+          THCudaCheck(cudaEventCreateWithFlags(&event, cudaEventDisableTiming));
+          THCudaCheck(cudaEventRecord(event, THCStream_stream(parent_stream)));
+          has_event = true;
+        }
+
+        // Has the child stream wait for the parent to be complete
+        THCudaCheck(cudaStreamWaitEvent(
+          THCStream_stream(child_metadata.stream())
+        , event
+        , 0));
+      }
+    #endif 
```

Comment:
Wow, this is complicated. This has to at least become a helper function, because it adds a ton of unnecessary noise to `evaluate_function`.

Reply:
I'll greatly simplify this code and (per your point later) make the event an RAII. I think it'll look really good afterwards.